presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -358,58 +245,110 @@ private static void writePolyline(Geometry geometry, SliceOutput output, boolean
+                writeMultiPoint((MultiPoint) geometry, output);
+                break;
+            case "LineString":
+            case "MultiLineString":
+                writePolyline(geometry, output);
+                break;
+            case "Polygon":
+            case "MultiPolygon":
+                writePolygon(geometry, output);
+                break;
+            case "GeometryCollection":
+                writeGeometryCollection(output, geometry);
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported geometry type : " + geometry.getGeometryType());
         }
+    }
 
-        int pointCount = input.readInt();
-        int[] startIndexes = new int[partCount];
-        for (int i = 0; i < partCount; i++) {
-            startIndexes[i] = input.readInt();
+    private static void writePoint(Point point, SliceOutput output)
+    {
+        output.writeByte(GeometryType.POINT.code());
+        output.writeInt(EsriShapeType.POINT.code);
+        if (!point.isEmpty()) {
+            writeCoordinate(point.getCoordinate(), output);
+        }
+        else {
+            writeCoordinate(new Coordinate(translateToAVNaN(Double.NaN), translateToAVNaN(Double.NaN)), output);
         }
+    }
 
-        int[] partLengths = new int[partCount];
-        if (partCount > 1) {
-            partLengths[0] = startIndexes[1];
-            for (int i = 1; i < partCount - 1; i++) {
-                partLengths[i] = startIndexes[i + 1] - startIndexes[i];
-            }
+    private static void writeMultiPoint(MultiPoint geometry, SliceOutput output)
+    {
+        output.writeByte(GeometryType.MULTI_POINT.code());
+        output.writeInt(EsriShapeType.MULTI_POINT.code);
+        writeEnvelope(geometry.getEnvelopeInternal(), output, geometry.isEmpty());
+        output.writeInt(geometry.getNumPoints());
+        for (Coordinate coordinate : geometry.getCoordinates()) {
+            writeCoordinate(coordinate, output);
         }
-        partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+    }
 
-        LinearRing shell = null;
-        List<LinearRing> holes = new ArrayList<>();
-        List<Polygon> polygons = new ArrayList<>();
-        for (int i = 0; i < partCount; i++) {
-            Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
-            if (!isClockwise(coordinates)) {
-                holes.add(GEOMETRY_FACTORY.createLinearRing(coordinates));
-                continue;
-            }
-            if (shell != null) {
-                polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
-            }
-            shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
-            holes.clear();
+    private static void writePolyline(Geometry geometry, SliceOutput output)
+    {
+        int numParts;
+        int numPoints = geometry.getNumPoints();
+        switch (geometry.getGeometryType()) {
```

Comment:
This `switch` happens 3 times in a row. First, in `writeGeometry` and then twice here. Better add `boolean multiType` argument to `writePolyline`.